File: lec08/01_hash_table_sc/hashTableSC.c
// hashTableSC.c - Hash Table with Separate Chaining
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// 해시 테이블 노드 구조체 (연결 리스트의 노드)
struct HashNode {
  int key;
  int value;
  struct HashNode* next;
};

// 해시 테이블 구조체
struct HashTable {
  int size;
  struct HashNode** table;
};

// 해시 함수 (간단한 modulo 연산)
int hashFunction(int key, int size) {
  return key % size;
}

// 새 노드 생성 함수
struct HashNode* createNode(int key, int value) {
  struct HashNode* newNode = (struct HashNode*)malloc(sizeof(struct HashNode));
  newNode->key = key;
  newNode->value = value;
  newNode->next = NULL;
  return newNode;
}

// 해시 테이블 생성 함수
struct HashTable* createHashTable(int size) {
  struct HashTable* hashTable =
      (struct HashTable*)malloc(sizeof(struct HashTable));
  hashTable->size = size;
  hashTable->table = (struct HashNode**)malloc(size * sizeof(struct HashNode*));

  // 모든 버킷을 NULL로 초기화
  for (int i = 0; i < size; i++) {
    hashTable->table[i] = NULL;
  }

  return hashTable;
}

// 삽입 함수
void insert(struct HashTable* hashTable, int key, int value) {
  int index = hashFunction(key, hashTable->size);
  struct HashNode* current = hashTable->table[index];

  // 기존 키가 있는지 확인 (업데이트)
  while (current != NULL) {
    if (current->key == key) {
      current->value = value;
      return;
    }
    current = current->next;
  }

  // 새 노드를 리스트 앞에 삽입
  struct HashNode* newNode = createNode(key, value);
  newNode->next = hashTable->table[index];
  hashTable->table[index] = newNode;
}

// 검색 함수
int* search(struct HashTable* hashTable, int key) {
  int index = hashFunction(key, hashTable->size);
  struct HashNode* current = hashTable->table[index];

  while (current != NULL) {
    if (current->key == key) {
      return &(current->value);
    }
    current = current->next;
  }

  return NULL;  // 키를 찾지 못함
}

// 삭제 함수
int deleteKey(struct HashTable* hashTable, int key) {
  int index = hashFunction(key, hashTable->size);
  struct HashNode* current = hashTable->table[index];
  struct HashNode* prev = NULL;

  while (current != NULL) {
    if (current->key == key) {
      if (prev == NULL) {
        // 첫 번째 노드 삭제
        hashTable->table[index] = current->next;
      } else {
        // 중간 또는 마지막 노드 삭제
        prev->next = current->next;
      }

      free(current);
      return 1;  // 삭제 성공
    }

    prev = current;
    current = current->next;
  }

  return 0;  // 키를 찾지 못함
}

// 해시 테이블 출력 함수
void printHashTable(struct HashTable* hashTable) {
  printf("해시 테이블 내용:\n");

  for (int i = 0; i < hashTable->size; i++) {
    printf("버킷 %d: ", i);
    struct HashNode* current = hashTable->table[i];

    if (current == NULL) {
      printf("(비어있음)");
    } else {
      while (current != NULL) {
        printf("[%d:%d]", current->key, current->value);
        if (current->next != NULL) {
          printf(" -> ");
        }
        current = current->next;
      }
    }
    printf("\n");
  }
}

// 해시 테이블 메모리 해제 함수
void freeHashTable(struct HashTable* hashTable) {
  for (int i = 0; i < hashTable->size; i++) {
    struct HashNode* current = hashTable->table[i];
    while (current != NULL) {
      struct HashNode* temp = current;
      current = current->next;
      free(temp);
    }
  }

  free(hashTable->table);
  free(hashTable);
}

// 해시 테이블 크기 조정 함수
struct HashTable* resize(struct HashTable* oldTable, int newSize) {
  struct HashTable* newTable = createHashTable(newSize);

  // 기존 모든 요소를 새 테이블에 재삽입
  for (int i = 0; i < oldTable->size; i++) {
    struct HashNode* current = oldTable->table[i];
    while (current != NULL) {
      insert(newTable, current->key, current->value);
      current = current->next;
    }
  }

  return newTable;
}

// 사용 예시
int main() {
  // 크기 7인 해시 테이블 생성
  struct HashTable* hashTable = createHashTable(7);

  printf("=== 해시 테이블 (Separate Chaining) 테스트 ===\n\n");

  // 데이터 삽입
  printf(
      "데이터 삽입: (10,100), (22,220), (31,310), (4,40), (15,150), (28,280), "
      "(17,170)\n");
  insert(hashTable, 10, 100);
  insert(hashTable, 22, 220);
  insert(hashTable, 31, 310);
  insert(hashTable, 4, 40);
  insert(hashTable, 15, 150);
  insert(hashTable, 28, 280);
  insert(hashTable, 17, 170);

  printHashTable(hashTable);

  // 검색 테스트
  printf("\n=== 검색 테스트 ===\n");
  int* result = search(hashTable, 22);
  if (result != NULL) {
    printf("키 22 검색 결과: %d\n", *result);
  } else {
    printf("키 22를 찾을 수 없습니다.\n");
  }

  result = search(hashTable, 99);
  if (result != NULL) {
    printf("키 99 검색 결과: %d\n", *result);
  } else {
    printf("키 99를 찾을 수 없습니다.\n");
  }

  // 업데이트 테스트
  printf("\n=== 업데이트 테스트 ===\n");
  printf("키 22의 값을 999로 업데이트\n");
  insert(hashTable, 22, 999);
  printHashTable(hashTable);

  // 삭제 테스트
  printf("\n=== 삭제 테스트 ===\n");
  printf("키 22 삭제\n");
  if (deleteKey(hashTable, 22)) {
    printf("키 22 삭제 성공\n");
  } else {
    printf("키 22 삭제 실패\n");
  }
  printHashTable(hashTable);

  printf("\n키 99 삭제 시도\n");
  if (deleteKey(hashTable, 99)) {
    printf("키 99 삭제 성공\n");
  } else {
    printf("키 99 삭제 실패 (존재하지 않음)\n");
  }

  // 크기 조정 테스트
  printf("\n=== 크기 조정 테스트 ===\n");
  printf("해시 테이블 크기를 7에서 13으로 조정\n");
  struct HashTable* resizedTable = resize(hashTable, 13);
  printf("조정된 해시 테이블:\n");
  printHashTable(resizedTable);

  // 메모리 해제
  freeHashTable(hashTable);
  freeHashTable(resizedTable);

  return 0;
}

File: lec08/01_hash_table_sc/hashTableSC.py
# hashTableSC.py - Hash Table with Separate Chaining

class HashNode:
  """해시 테이블 노드 클래스 (연결 리스트의 노드)"""

  def __init__(self, key, value):
    self.key = key
    self.value = value
    self.next = None


class HashTableSC:
  """Separate Chaining 방식의 해시 테이블 클래스"""

  def __init__(self, size=7):
    """해시 테이블 초기화"""
    self.size = size
    self.table = [None] * size

  def _hash_function(self, key):
    """해시 함수 (간단한 modulo 연산)"""
    return key % self.size

  def insert(self, key, value):
    """키-값 쌍을 해시 테이블에 삽입"""
    index = self._hash_function(key)
    current = self.table[index]

    # 기존 키가 있는지 확인 (업데이트)
    while current is not None:
      if current.key == key:
        current.value = value
        return
      current = current.next

    # 새 노드를 리스트 앞에 삽입
    new_node = HashNode(key, value)
    new_node.next = self.table[index]
    self.table[index] = new_node

  def search(self, key):
    """키에 해당하는 값을 검색"""
    index = self._hash_function(key)
    current = self.table[index]

    while current is not None:
      if current.key == key:
        return current.value
      current = current.next

    return None  # 키를 찾지 못함

  def delete(self, key):
    """키를 해시 테이블에서 삭제"""
    index = self._hash_function(key)
    current = self.table[index]
    prev = None

    while current is not None:
      if current.key == key:
        if prev is None:
          # 첫 번째 노드 삭제
          self.table[index] = current.next
        else:
          # 중간 또는 마지막 노드 삭제
          prev.next = current.next
        return True  # 삭제 성공

      prev = current
      current = current.next

    return False  # 키를 찾지 못함

  def print_table(self):
    """해시 테이블 내용을 출력"""
    print("해시 테이블 내용:")

    for i in range(self.size):
      print(f"버킷 {i}: ", end="")
      current = self.table[i]

      if current is None:
        print("(비어있음)")
      else:
        chain = []
        while current is not None:
          chain.append(f"[{current.key}:{current.value}]")
          current = current.next
        print(" -> ".join(chain))

  def get_all_items(self):
    """모든 키-값 쌍을 리스트로 반환"""
    items = []
    for i in range(self.size):
      current = self.table[i]
      while current is not None:
        items.append((current.key, current.value))
        current = current.next
    return items

  def resize(self, new_size):
    """해시 테이블 크기 조정"""
    # 기존 모든 아이템을 저장
    old_items = self.get_all_items()

    # 새로운 크기로 테이블 초기화
    self.size = new_size
    self.table = [None] * new_size

    # 모든 아이템을 새 테이블에 재삽입
    for key, value in old_items:
      self.insert(key, value)

  def get_load_factor(self):
    """로드 팩터 계산"""
    item_count = len(self.get_all_items())
    return item_count / self.size

  def get_statistics(self):
    """해시 테이블 통계 정보 반환"""
    non_empty_buckets = 0
    max_chain_length = 0
    total_chain_length = 0

    for i in range(self.size):
      chain_length = 0
      current = self.table[i]

      if current is not None:
        non_empty_buckets += 1
        while current is not None:
          chain_length += 1
          current = current.next

      max_chain_length = max(max_chain_length, chain_length)
      total_chain_length += chain_length

    return {
        "총 버킷 수": self.size,
        "사용된 버킷 수": non_empty_buckets,
        "총 아이템 수": total_chain_length,
        "로드 팩터": self.get_load_factor(),
        "최대 체인 길이": max_chain_length,
        "평균 체인 길이": total_chain_length / self.size if self.size > 0 else 0
    }


# 사용 예시
if __name__ == "__main__":
  # 크기 7인 해시 테이블 생성
  hash_table = HashTableSC(7)

  print("=== 해시 테이블 (Separate Chaining) 테스트 ===\n")

  # 데이터 삽입
  print("데이터 삽입: (10,100), (22,220), (31,310), (4,40), (15,150), (28,280), (17,170)")
  items_to_insert = [(10, 100), (22, 220), (31, 310), (4, 40),
                     (15, 150), (28, 280), (17, 170)]

  for key, value in items_to_insert:
    hash_table.insert(key, value)

  hash_table.print_table()

  # 검색 테스트
  print("\n=== 검색 테스트 ===")
  result = hash_table.search(22)
  if result is not None:
    print(f"키 22 검색 결과: {result}")
  else:
    print("키 22를 찾을 수 없습니다.")

  result = hash_table.search(99)
  if result is not None:
    print(f"키 99 검색 결과: {result}")
  else:
    print("키 99를 찾을 수 없습니다.")

  # 업데이트 테스트
  print("\n=== 업데이트 테스트 ===")
  print("키 22의 값을 999로 업데이트")
  hash_table.insert(22, 999)
  hash_table.print_table()

  # 삭제 테스트
  print("\n=== 삭제 테스트 ===")
  print("키 22 삭제")
  if hash_table.delete(22):
    print("키 22 삭제 성공")
  else:
    print("키 22 삭제 실패")
  hash_table.print_table()

  print("\n키 99 삭제 시도")
  if hash_table.delete(99):
    print("키 99 삭제 성공")
  else:
    print("키 99 삭제 실패 (존재하지 않음)")

  # 통계 정보
  print("\n=== 해시 테이블 통계 ===")
  stats = hash_table.get_statistics()
  for key, value in stats.items():
    if isinstance(value, float):
      print(f"{key}: {value:.2f}")
    else:
      print(f"{key}: {value}")

  # 크기 조정 테스트
  print("\n=== 크기 조정 테스트 ===")
  print("해시 테이블 크기를 7에서 13으로 조정")
  hash_table.resize(13)
  print("조정된 해시 테이블:")
  hash_table.print_table()

  print("\n조정 후 통계:")
  stats = hash_table.get_statistics()
  for key, value in stats.items():
    if isinstance(value, float):
      print(f"{key}: {value:.2f}")
    else:
      print(f"{key}: {value}")
